backend/ipip/ipip.go

```diff
@@ -0,0 +1,182 @@
+}
+
+func New(sm subnet.Manager, extIface *backend.ExternalInterface) (backend.Backend, error) {
+	be := &IPIPBackend{
+		sm:       sm,
+		extIface: extIface,
+	}
+	return be, nil
+}
+
+func (be *IPIPBackend) RegisterNetwork(ctx context.Context, config *subnet.Config) (backend.Network, error) {
+	cfg := struct {
+		DirectRouting bool
+	}{}
+	if len(config.Backend) > 0 {
+		if err := json.Unmarshal(config.Backend, &cfg); err != nil {
+			return nil, fmt.Errorf("error decoding IPIP backend config: %v", err)
+		}
+	}
+	glog.Infof("IPIP config: DirectRouting=%v", cfg.DirectRouting)
+
+	n := &network{
+		SimpleNetwork: backend.SimpleNetwork{
+			ExtIface: be.extIface,
+		},
+		sm:          be.sm,
+		backendType: backendType,
+	}
+
+	attrs := &subnet.LeaseAttrs{
+		PublicIP:    ip.FromIP(be.extIface.ExtAddr),
+		BackendType: backendType,
+	}
+
+	l, err := be.sm.AcquireLease(ctx, attrs)
+	switch err {
+	case nil:
+		n.SubnetLease = l
+	case context.Canceled, context.DeadlineExceeded:
+		return nil, err
+	default:
+		return nil, fmt.Errorf("failed to acquire lease: %v", err)
+	}
+	dev, err := be.configureIPIPDevice(n.SubnetLease)
+	if err != nil {
+		return nil, err
+	}
+	dev.directRouting = cfg.DirectRouting
+	n.dev = dev
+
+	return n, nil
+}
+
+func (be *IPIPBackend) configureIPIPDevice(lease *subnet.Lease) (*tunnelDev, error) {
+	link, err := netlink.LinkByName(tunnelName)
+	if err != nil {
+		// When modprobe ipip module, a tunl0 ipip device is created automatically per network namespace by ipip kernel module.
+		// It is the namespace default IPIP device with attributes local=any and remote=any.
+		// When receiving IPIP protocol packets, kernel will forward them to tunl0 as a fallback device
+		// if it can't find an option whose local/remote attribute matches their src/dst ip address more precisely.
+		// See https://github.com/torvalds/linux/blob/v4.13/net/ipv4/ip_tunnel.c#L85-L95 .
+
+		// So we have two options of creating ipip device, either rename tunl0 to flannel.ipip or create an new ipip device
+		// and set local attribute of flannel.ipip to distinguish these two devices.
+		// Considering tunl0 might be used by users, so choose the later option.
+		link = &netlink.Iptun{LinkAttrs: netlink.LinkAttrs{Name: tunnelName}, Local: be.extIface.IfaceAddr}
+		if err := netlink.LinkAdd(link); err != nil {
+			return nil, fmt.Errorf("failed to create tunnel %v: %v", tunnelName, err)
+		}
+	}
+	// flannel will never make this happen. This situation can only be caused by a user, so get them to sort it out.
+	if link.Type() != "ipip" {
+		return nil, fmt.Errorf("%v isn't an ipip mode device", tunnelName)
+	}
+	ipip, ok := link.(*netlink.Iptun)
+	if !ok {
+		return nil, fmt.Errorf("%s isn't an iptun device %#v", tunnelName, link)
+	}
+
+	// Don't set remote attribute making flannel.ipip an one to many tunnel device.
+	if (ipip.Local != nil && !ipip.Local.Equal(be.extIface.IfaceAddr)) || (ipip.Remote != nil && ipip.Remote.String() != "0.0.0.0") {
```

Comment:
I don't think that just returning an error here is good enough. If there's a `flannel.ipip` that isn't an ipip device then it wasn't created by flannel, so in that case it's best to just return an error. But in this case, the local address can change between flannel restarts, so flannel should be able to recreate the device and not just return an error. You should follow similar logic to vxlan (see the ensureLink function), it should delete and recreate the device if the config changes

Reply:
Fixed